plugins/flatpak/gs-flatpak.c

```diff
@@ -2502,6 +2502,84 @@ install_runtime_for_app (GsFlatpak *self,
 	return TRUE;
 }
 
+static gboolean
+gs_flatpak_app_source_equals_remote (GsApp *source, FlatpakRemote *remote)
+{
+	const gchar *source_url = gs_app_get_metadata_item (source, "flatpak::url");
+	g_autofree gchar *remote_url = flatpak_remote_get_url (remote);
+	return g_strcmp0 (source_url, remote_url) == 0;
+}
+
+static FlatpakRemote *
+gs_flatpak_get_remote_for_source (GsFlatpak *self,
+				  GsApp *app,
+				  gboolean *same_name_remote_exists,
+				  GCancellable *cancellable)
+{
+	FlatpakRemote *remote = NULL;
+	g_autoptr(GPtrArray) remotes = NULL;
+
+	/* check an eventual remote with the same name */
+	remote = flatpak_installation_get_remote_by_name (self->installation,
+							  gs_app_get_id (app),
+							  cancellable, NULL);
+
+	if (remote != NULL) {
+		if (gs_flatpak_app_source_equals_remote (app, remote))
+			return remote;
+		*same_name_remote_exists = TRUE;
+		return NULL;
+	}
+
+	remotes = flatpak_installation_list_remotes (self->installation,
+						     cancellable,
+						     NULL);
+	if (remotes == NULL) {
+		return NULL;
+	}
+
+	for (guint i = 0; i < remotes->len; ++i) {
+		remote = g_ptr_array_index (remotes, i);
+		if (gs_flatpak_app_source_equals_remote (app, remote))
+			return remote;
```

Comment:
Don't you want to take a reference to the remote? Otherwise when the ptr array goes out of scope, you lose the only reference that is owned by this function